From 1cd8a2de1baabdaa2d908acc76b1743561a44bd1 Mon Sep 17 00:00:00 2001
From: Hyun Kwon <hyun.kwon@xilinx.com>
Date: Wed, 22 Jul 2020 09:55:44 -0700
Subject: [PATCH 1463/1851] misc: xilinx-ai-engine: remove duplicate module
 init/exit

commit 262fadfa6ca922191ce9b365196ddbd41764cbcf from
https://github.com/Xilinx/linux-xlnx.git

module_platform_driver() simply translates into module_init() and
module_exit(), hence conflicts as in below snippet. Fix it by moving
platform driver registration into existing module init/exit.

In file included from drivers/misc/xilinx-ai-engine/ai-engine-dev.c:17:
./include/linux/module.h:128:42: error: redefinition of '__inittest'
  128 |  static inline initcall_t __maybe_unused __inittest(void)  \
      |                                          ^~~~~~~~~~
./include/linux/device.h:1903:1: note: in expansion of macro 'module_init'
 1903 | module_init(__driver##_init); \
      | ^~~~~~~~~~~
./include/linux/platform_device.h:238:2: note: in expansion of macro 'module_driver'
  238 |  module_driver(__platform_driver, platform_driver_register, \
      |  ^~~~~~~~~~~~~
drivers/misc/xilinx-ai-engine/ai-engine-dev.c:444:1: note: in expansion of macro 'module_platform_driver'
  444 | module_platform_driver(xilinx_ai_engine_driver);
      | ^~~~~~~~~~~~~~~~~~~~~~
./include/linux/module.h:128:42: note: previous definition of '__inittest' was here
  128 |  static inline initcall_t __maybe_unused __inittest(void)  \
      |                                          ^~~~~~~~~~
./include/linux/module.h:111:32: note: in expansion of macro 'module_init'
  111 | #define postcore_initcall(fn)  module_init(fn)
      |                                ^~~~~~~~~~~
drivers/misc/xilinx-ai-engine/ai-engine-dev.c:435:1: note: in expansion of macro 'postcore_initcall'
  435 | postcore_initcall(xilinx_ai_engine_init);
      | ^~~~~~~~~~~~~~~~~
./include/linux/module.h:130:6: error: redefinition of 'init_module'
...

Reported-by: kernel test robot <lkp@intel.com>
Signed-off-by: Hyun Kwon <hyun.kwon@xilinx.com>
Tested-by: Wendy Liang <wendy.liang@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/misc/xilinx-ai-engine/ai-engine-dev.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/misc/xilinx-ai-engine/ai-engine-dev.c b/drivers/misc/xilinx-ai-engine/ai-engine-dev.c
index 048a4a4c4894..042c52372bb3 100644
--- a/drivers/misc/xilinx-ai-engine/ai-engine-dev.c
+++ b/drivers/misc/xilinx-ai-engine/ai-engine-dev.c
@@ -591,18 +591,19 @@ static int __init xilinx_ai_engine_init(void)
 		return PTR_ERR(aie_class);
 	}
 
+	platform_driver_register(&xilinx_ai_engine_driver);
+
 	return 0;
 }
 postcore_initcall(xilinx_ai_engine_init);
 
 static void __exit xilinx_ai_engine_exit(void)
 {
+	platform_driver_unregister(&xilinx_ai_engine_driver);
 	class_destroy(aie_class);
 	unregister_chrdev_region(aie_major, AIE_DEV_MAX);
 }
 module_exit(xilinx_ai_engine_exit);
 
-module_platform_driver(xilinx_ai_engine_driver);
-
 MODULE_AUTHOR("Xilinx, Inc.");
 MODULE_LICENSE("GPL v2");
-- 
2.31.1

